pkg/cni/types.go

```diff
@@ -0,0 +1,104 @@
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/k8sclient"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/types"
+	"net/http"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+type cniRequestFunc func(request *PodRequest) ([]byte, error)
+
+// Explicit type for CNI commands the server handles
+type command string
+
+// PodRequest represents a request for ADD / DEL / CHECK for a Pod
+type PodRequest struct {
+	// The CNI command of the operation
+	Command command
+	// kubernetes namespace name
+	PodNamespace string
+	// kubernetes pod name
+	PodName string
+	// kubernetes pod UID
+	PodUID string
+	// kubernetes container ID
+	SandboxID string
+	// CNI container ID
+	ContainerID string
+	// kernel network namespace path
+	Netns string
+	// Interface name to be configured
+	IfName string
+	// CNI conf obtained from stdin conf
+	CNIConf *types.NetConf
+	// Timestamp when the request was started
+	timestamp time.Time
+	// ctx is a context tracking this request's lifetime
+	ctx context.Context
+	// cancel should be called to cancel this request
+	cancel context.CancelFunc
+	// kubeclient has the kubernetes API client
+	kubeclient *k8sclient.ClientInfo
+	// exec execs
+	exec invoke.Exec
+}
+
+// Request sent to the Server by the multus-shim
+type Request struct {
+	// CNI environment variables, like CNI_COMMAND and CNI_NETNS
+	Env map[string]string `json:"env,omitempty"`
+	// CNI configuration passed via stdin to the CNI plugin
+	Config []byte `json:"config,omitempty"`
+}
+
+// KubeAPIAuth contains information necessary to create a Kube API client
+type KubeAPIAuth struct {
+	// Kubeconfig is the path to a kubeconfig
+	Kubeconfig string `json:"kubeconfig,omitempty"`
+	// KubeAPIServer is the URL of a Kubernetes API server (not required if kubeconfig is given)
+	KubeAPIServer string `json:"kube-api-server,omitempty"`
+	// KubeAPIToken is a Kubernetes API token (not required if kubeconfig is given)
+	KubeAPIToken string `json:"kube-api-token,omitempty"`
+	// KubeCAData is the Base64-ed Kubernetes API CA certificate data (not required if kubeconfig is given)
+	KubeCAData string `json:"kube-ca-data,omitempty"`
+}
+
+// Server represents an HTTP server listening to a unix socket. It will handle
+// the CNI shim requests issued when a pod is added / removed.
+type Server struct {
+	http.Server
+	requestFunc       cniRequestFunc
+	rundir            string
+	useOVSExternalIDs int32
+	kubeAuth          *KubeAPIAuth
+	kubeclient        *k8sclient.ClientInfo
+	exec              invoke.Exec
+}
+
+// CNIAdd is the command representing add operation for a new pod
+const CNIAdd command = "ADD"
+
+// CNIUpdate is the command representing update operation for an existing pod
+const CNIUpdate command = "UPDATE"
+
+// CNIDel is the command representing delete operation on a pod that is to be torn down
+const CNIDel command = "DEL"
+
+// CNICheck is the command representing check operation on a pod
+const CNICheck command = "CHECK"
```

Comment:
I suppose `const CNIAdd` seems to be redundant. We just use 'ADD' string, not make it const variable.

Reply:
I guess I'd prefer to have more uses of the CNIAdd const rather than call `"ADD"`; DRY principle. 

But I'll follow your suggestion if you insist.

Reply:
But to prevent to repeat, adding more code is not make sense to me.
"ADD" is 5 chars, CNIAdd is 6 chars. In this case, I suppose we could repeat it.

Reply:
I don't think the important parameter is the number of characters - I just think that using magic strings is worse than using a const, since I can update the value of the const once rather than searching the code N times for the repetitions of the magic string. 

I'm aware it isn't probably that this string will ever be updated, but still, as a practice, I'd take a const over magic strings every time.

I'll remove the consts and add more magic strings, since you insist.

Reply:
Done.